File: variants/haskell/src/RCA/Control.hs
-------------------------------------------------------------------------------
-- Regulated Cell Architecture (RCA) - Haskell Variant
-- Control Plane
-------------------------------------------------------------------------------

module RCA.Control where

-------------------------------------------------------------------------------
-- (1) States
-------------------------------------------------------------------------------

-- Status: FREEZE
data State = Init | Idle | Running | Halt | Failure | Shutdown
  deriving (Show, Eq)

-------------------------------------------------------------------------------
-- (2) Modes
-------------------------------------------------------------------------------

-- Status: MUTABLE
data Mode = ModeNone | Debug
  deriving (Show, Eq)

-------------------------------------------------------------------------------
-- (3) Events
-------------------------------------------------------------------------------

-- Status: MUTABLE
data Event = EventNone
  deriving (Show, Eq)

-------------------------------------------------------------------------------
-- Control Plane
-------------------------------------------------------------------------------

-- Status: FREEZE
data ControlPlane = ControlPlane
  { cpState :: State
  , cpMode  :: Mode
  , cpEvent :: Event
  } deriving (Show)

defaultControlPlane :: ControlPlane
defaultControlPlane = ControlPlane
  { cpState = Init
  , cpMode  = Debug
  , cpEvent = EventNone
  }

File: variants/python/rca/control.py
"""
Regulated Cell Architecture (RCA) - Python Variant
Control Plane
"""

from __future__ import annotations
from dataclasses import dataclass
from enum import Enum, auto


###############################################################################
# (1) States
###############################################################################

# Status: FREEZE
class State(Enum):
    INIT = auto()
    IDLE = auto()
    RUNNING = auto()
    HALT = auto()
    FAILURE = auto()
    SHUTDOWN = auto()


###############################################################################
# (2) Modes
###############################################################################

# Status: MUTABLE
class Mode(Enum):
    NONE = auto()
    DEBUG = auto()


###############################################################################
# (3) Events
###############################################################################

# Status: MUTABLE
class Event(Enum):
    NONE = auto()


###############################################################################
# Control Plane
###############################################################################

# Status: FREEZE
@dataclass
class ControlPlane:
    state: State = State.INIT
    mode: Mode = Mode.DEBUG
    event: Event = Event.NONE
